core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildRemoteContainerRunner.java

```diff
@@ -34,59 +39,93 @@ protected void preBuild(Path outputDir, List<String> buildArgs) throws Interrupt
-            throw new RuntimeException(errorMsg);
+            throw new RuntimeException("Command failed: " + String.join(" ", command));
         }
         try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
-            return reader.readLine();
+            return reader.lines().toList();
         }
     }
 
     @Override
     protected void postBuild(Path outputDir, String nativeImageName, String resultingExecutableName) {
         copyFromContainerVolume(outputDir, resultingExecutableName,
                 "Failed to copy native image from container volume back to the host.");
+
+        // Note that podman cp does not support globbing i.e. cp /project/*.so will not work.
+        // Why only .so? How about .dynlib and .lib? Regardless of the host platform,
+        // the builder container is always Linux. So, we only need to copy .so files.
+        //
+        // We could either start the container again, exec `find' or `ls' to list the .so files,
+        // stop the container again and use that list. We could also use the build-artifacts.json
+        // to get the list of artifacts straight away which is what ended up doing here:
+        copyFromContainerVolume(outputDir, "build-artifacts.json", null);
+        try {
+            final Path buildArtifactsFile = outputDir.resolve("build-artifacts.json");
+            if (Files.exists(buildArtifactsFile)) {
+                // The file is small enough to afford this read
+                final String buildArtifactsJson = Files.readString(buildArtifactsFile);
+                final Matcher m = JDK_LIBS.matcher(buildArtifactsJson);
+                if (m.matches()) {
+                    for (String lib : m.group(1).split(",")) {
+                        final String l = lib.trim().replace("\"", "");
+                        copyFromContainerVolume(outputDir, l,
+                                "Failed to copy " + l + " from container volume back to the host.");
+                    }
+                }
+            }
+        } catch (IOException e) {
+            log.errorf(e, "Failed to list .so files in the build-artifacts.json. Skipping the step.");
+        }
+
         if (nativeConfig.debug().enabled()) {
-            copyFromContainerVolume(outputDir, "sources", "Failed to copy sources from container volume back to the host.");
-            String symbols = String.format("%s.debug", nativeImageName);
-            copyFromContainerVolume(outputDir, symbols, "Failed to copy debug symbols from container volume back to the host.");
+            copyFromContainerVolume(outputDir, "sources",
+                    "Failed to copy sources from container volume back to the host.");
+            final String symbols = String.format("%s.debug", nativeImageName);
+            copyFromContainerVolume(outputDir, symbols,
+                    "Failed to copy debug symbols from container volume back to the host.");
```

Comment:
Are these changes necessary?

Reply:
These changes are not necessary to fix the issue.